sdk/core/System.ClientModel/src/ModelReaderWriter/JsonCollectionWriter.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections;
+using System.Text.Json;
+
+namespace System.ClientModel.Primitives;
+
+internal class JsonCollectionWriter : CollectionWriter
+{
+    internal override BinaryData Write(IEnumerable enumerable, ModelReaderWriterOptions options)
+    {
+        using UnsafeBufferSequence sequenceWriter = new();
+        using Utf8JsonWriter writer = new(sequenceWriter);
+        WriteEnumerable(enumerable, writer, options);
+        writer.Flush();
+        return sequenceWriter.ExtractReader().ToBinaryData();
+    }
+
+    private static void WriteJson(object model, Utf8JsonWriter writer, ModelReaderWriterOptions options)
+    {
+        if (model is IJsonModel<object> jsonModel)
+        {
+            jsonModel.Write(writer, options);
+        }
+        else if (model is IEnumerable enumerable)
+        {
+            WriteEnumerable(enumerable, writer, options);
+        }
+        else
+        {
+            throw new InvalidOperationException($"{model.GetType().Name} does not implement IJsonModel or IEnumerable<IJsonModel>");
+        }
+    }
+
+    private static void WriteEnumerable(IEnumerable enumerable, Utf8JsonWriter writer, ModelReaderWriterOptions options)
+    {
+        var enumerableType = enumerable.GetType();
+
+        if (enumerableType.IsArray &&
+            enumerableType.GetArrayRank() > 1 &&
+            enumerableType.GetElementType()?.IsArray == false) //multi-dimensional array
+        {
+            Array array = (Array)enumerable;
+            WriteMultiDimensionalArray(array, new int[array.Rank], 0, options, writer);
+        }
+        else
+        {
+            if (enumerable is IDictionary dictionary)
+            {
+                writer.WriteStartObject();
+                foreach (var key in dictionary.Keys)
+                {
+                    writer.WritePropertyName(key.ToString()!);
+                    WriteJson(dictionary[key]!, writer, options);
+                }
+                writer.WriteEndObject();
+            }
+            else
+            {
+                writer.WriteStartArray();
+                foreach (var item in enumerable)
+                {
+                    WriteJson(item, writer, options);
+                }
+                writer.WriteEndArray();
+            }
+        }
+    }
+
+    private static void WriteMultiDimensionalArray(Array array, int[] indices, int currentDimension, ModelReaderWriterOptions options, Utf8JsonWriter writer)
```

Comment:
should we just throw for multi-dim arrays?

Reply:
We could throw, but this does work today we have test cases for both [,] and [][] and it creates the write output in both cases.